### PR TITLE
Make supervisor API similar to submit task API

### LINKED ISSUE #8662: Normalize the submit task API to be more like the supervisor API
### Description

Right now submitting a task expects a payload like:

```
{
  "type": "index",
  "spec": {
    "dataSchema": { ... },
    "tuningConfig": { ... },
    "ioConfig": { ... }
  }
}
```

(see https://druid.apache.org/docs/latest/tutorials/tutorial-batch.html#loading-data-with-a-spec-via-console)

While a supervisor has a payload of

```
{
  "type": "kafka",
  "dataSchema": { ... },
  "tuningConfig": { ... },
  "ioConfig": { ... }
}
```

(see https://druid.apache.org/docs/latest/tutorials/tutorial-kafka.html)

This is a hastle for the console to handle and also confusing to users ( https://stackoverflow.com/questions/58097806/cant-create-druid-ingestion-task-through-api/58131519#58131519 )

I suggest to normalize the task API to be able to accept 

```
{
  "type": "index",
  "dataSchema": { ... },
  "tuningConfig": { ... },
  "ioConfig": { ... }
}
```

as the payload. It should also be backwards compatible.

Implementation wise it could just rewrite it to the `spec` form internally.

### PR DESCRIPTION
**Pull request by surekhasaharan**
Fixes #8662

### Description

Make submit task and supervisor API's similar in structure. supervisor API accepts both forms:
```
{
  "type": "kafka",
  "spec": {
    "dataSchema": { ... },
    "tuningConfig": { ... },
    "ioConfig": { ... }
  }
}
```
and
```
{
  "type": "kafka",
   "dataSchema": { ... },
   "tuningConfig": { ... },
   "ioConfig": { ... }
}
```
The supervisor spec would allow both formats, but the serialization is done to the old format to allow for backwards compatibility.

This PR has:
- [x] been self-reviewed.
- [x] added unit tests or modified existing tests to cover new code paths.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by jihoonson** [comment]
Please fix LGTM alert.

**Comment by surekhasaharan** [comment]
> Please fix LGTM alert.

Yeah, i looked at that, it thinks `dataSchema` can be null [here](https://github.com/apache/incubator-druid/pull/8810/files#diff-0ce554adcd07cc3ea613b457eb800164R208) because of test cases, where null is passed for `dataSchema`, though those tests pass a valid `ingestionSchema`, and it's already checked in [here](https://github.com/apache/incubator-druid/pull/8810/files#diff-0ce554adcd07cc3ea613b457eb800164R133). Not sure how to make lgtm ignore that or fix otherwise.

**Comment by vogievetsky** [comment]
I am a strong 👍 on the design (I am biased since I filed the original issue).
This will allow us to take a nasty hack out of the web console.

**Comment by surekhasaharan** [comment]
> > Yeah, i looked at that, it thinks `dataSchema` can be null [here](https://github.com/apache/incubator-druid/pull/8810/files#diff-0ce554adcd07cc3ea613b457eb800164R208) because of test cases, where null is passed for `dataSchema`, though those tests pass a valid `ingestionSchema`, and it's already checked in [here](https://github.com/apache/incubator-druid/pull/8810/files#diff-0ce554adcd07cc3ea613b457eb800164R133). Not sure how to make lgtm ignore that or fix otherwise.
> 
> You can fix it by making a non-null `IndexIngestionSpec` first and then using it in the constructor.
> 
> Also, this PR changes only `IndexTask`. Are you going to change `ParallelIndexSupervisorTask` in a follow-up pr?

Yes, want to change `ParallelIndexSupervisorTask` as well in this PR, working on it.

**Comment by surekhasaharan** [comment]
Thanks @jihoonson for review and suggestion, I have updated this PR to allow supervisor schema to take in `spec` (the old format is still valid) similar to `index`, `index_parallel` and `hadoop` submit task APIs

**Comment by jihoonson** [comment]
@surekhasaharan thanks for updating the PR. Would you please merge master so that LGTM can be passed?

File: extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java [hunk 1]
```diff
@@ -131,6 +120,13 @@ public Supervisor createSupervisor()
     );
   }
 
+  @Override
+  @JsonProperty
+  public DataSchema getDataSchema()
+  {
+    return super.getDataSchema();
```
Comment (jihoonson) [inline extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java]: Seems unnecessary.

File: extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java [hunk 2]
```diff
@@ -131,6 +120,13 @@ public Supervisor createSupervisor()
...
+  }
+
   @Override
   @JsonProperty
   public KafkaSupervisorTuningConfig getTuningConfig()
```
Comment (jihoonson) [inline extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java]: This and `getIoConfig()` should be deprecated now.

File: extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java (hunk 1 above)
Comment (jihoonson) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java]: And `getDataSchema()` should be deprecated.

File: extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java [hunk 3]
```diff
@@ -145,10 +141,18 @@ public KafkaSupervisorIOConfig getIoConfig()
     return (KafkaSupervisorIOConfig) super.getIoConfig();
   }
 
+  @Override
+  @JsonProperty
+  public KafkaSupervisorIngestionSpec getSpec()
+  {
+    return (KafkaSupervisorIngestionSpec) super.getSpec();
```
Comment (jihoonson) [inline extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java]: `@Nullable`

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorSpec.java [hunk 4]
```diff
@@ -50,6 +50,7 @@
   protected final SeekableStreamIndexTaskClientFactory indexTaskClientFactory;
   protected final ObjectMapper mapper;
   protected final RowIngestionMetersFactory rowIngestionMetersFactory;
+  private final SeekableStreamSupervisorIngestionSpec ingestionSchema;
   private final DataSchema dataSchema;
```
Comment (jihoonson) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorSpec.java]: This class now has duplicate information between `ingestionSchema` and `dataSchema`, `ioConfig`, and `tuningConfig`. I suggest to remove `dataSchema`, `ioConfig`, and `tuningConfig` and keep only `ingestionSchema`.

File: extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java (hunk 1 above)
Comment (surekhasaharan) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java]: removed from subclass and deprecated in base class

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorSpec.java (hunk 4 above)
Comment (surekhasaharan) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorSpec.java]: These will still be required in subclasses `KafkaSupervisorSpec` and `KinesisSupervisorSpec` for backwards compatibility, so there will be some duplication.

Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorSpec.java]: It doesn't have to be duplicate. You can always create ingestionSchema and get things from it.

Comment (surekhasaharan) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorSpec.java]: ok right, removed the extra configs from this class.

File: extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java (hunk 3 above)
Comment (jihoonson) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java]: Hmm sorry now it's not nullable.

File: extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisorSpec.java [hunk 5]
```diff
@@ -158,23 +132,34 @@ public String toString()
...
   @Override
+  @Deprecated
   @JsonProperty
   public KinesisSupervisorIOConfig getIoConfig()
   {
     return (KinesisSupervisorIOConfig) super.getIoConfig();
   }
 
+  @Override
+  @Nullable
```
Comment (jihoonson) [inline extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisorSpec.java]: Same here. It's not nullable now.

**Comment by jihoonson** [comment]
Thanks. I left a couple of trivial comments. LGTM after removing wrong nullable annotations.

File: extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java (hunk 3 above)
Comment (surekhasaharan) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisorSpec.java]: oh, yeah removed

**Pull request merged**